### ISSUE TITLE
A two-process test harness, so distributed code paths can be proven on one machine

### ISSUE THREAD

**Issue by YeapJieShen** [YeapJieShen/single-image-super-resolution#166]
## Context

No test in the suite runs more than one process, so nothing exercises a distributed code path. Any
rank-sensitive change therefore ships unproven, which conflicts with the standing rule that a
correctness fix arrives with a test that was red before it.

## What this is

A test harness that runs a real two-process trainer using the gloo backend on CPU:

```python
Trainer(accelerator="cpu", devices=2, strategy="ddp")
```

`gloo` is available in this environment (checked), and the suite already runs CPU-backed integration
tests, so the pattern exists.

## What it can prove

- A logged metric reduces to a global value rather than one process's shard
- A file is written by one process rather than all of them
- A dataset is or is not split across processes
- A collective called by one process and not another fails, rather than hanging silently in production

## What it cannot prove

NCCL-specific behaviour, device placement on real GPUs, memory, or multi-GPU throughput. Those need
actual multi-GPU hardware. Containers do not help — they share the host's devices, so two containers
on one GPU is still one GPU.

## Why it comes first

It is the enabler for every rank-sensitive change in this arc. With it in place, a rank guard can be
proven at the moment it is written, instead of being written once and proven later.

## Acceptance

- [ ] A reusable fixture runs a two-process trainer on CPU
- [ ] It is stable across repeated runs, not green once
